src/sink/otel.rs

```diff
@@ -0,0 +1,194 @@
+            Ok(SinkManager::new_box(Self {
+                config,
+                endpoint,
+                remote: None,
+            }))
+        } else {
+            Err("Offramp otel requires a config".into())
+        }
+    }
+}
+
+fn json_otel_logs_to_pb(json: &Value<'_>) -> Result<ExportLogsServiceRequest> {
+    let pb = ExportLogsServiceRequest {
+        resource_logs: logs::resource_logs_to_pb(json)?,
+    };
+    Ok(pb)
+}
+
+fn json_otel_trace_to_pb(json: &Value<'_>) -> Result<ExportTraceServiceRequest> {
+    let pb = ExportTraceServiceRequest {
+        resource_spans: trace::resource_spans_to_pb(Some(json))?,
+    };
+    Ok(pb)
+}
+
+fn json_otel_metrics_to_pb(json: &Value<'_>) -> Result<ExportMetricsServiceRequest> {
+    let pb = ExportMetricsServiceRequest {
+        resource_metrics: metrics::resource_metrics_to_pb(Some(json))?,
+    };
+    Ok(pb)
+}
+
+#[async_trait::async_trait]
+impl Sink for OpenTelemetry {
+    async fn terminate(&mut self) {}
+
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        _codec: &mut dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        mut event: Event,
+    ) -> ResultVec {
+        if let Some(_remote) = &mut self.remote {
+            for value in event.value_iter() {
+                if let Some(ref mut remote) = &mut self.remote {
+                    if let Value::Object(o) = value {
+                        if o.contains_key("metrics") {
+                            let request = json_otel_metrics_to_pb(value)?;
+                            match remote.metrics_client.export(request).await {
+                                Ok(_response) => {}
+                                Err(_e) => {
+                                    // TODO enhance error handling
+                                    error!("Failed to dispatch otel/gRPC metrics message");
+                                }
+                            };
+                        } else if o.contains_key("logs") {
+                            let request = json_otel_logs_to_pb(value)?;
+                            match remote.logs_client.export(request).await {
+                                Ok(_response) => {}
+                                Err(e) => {
+                                    error!("Failed to dispatch otel/gRPC log message, {}", e);
+                                }
+                            };
+                        } else if o.contains_key("trace") {
+                            let request = json_otel_trace_to_pb(value)?;
+                            match remote.trace_client.export(request).await {
+                                Ok(_response) => {}
+                                Err(e) => {
+                                    error!("Failed to dispatch otel/gRPC trace message, {}", e);
+                                }
+                            };
+                        } else {
+                            // unknown / drop
```

Comment:
what is the intended logic here? Should we log with `debug` or `trace` level?

Reply:
Vestigial leftover from debugging / testing with various otel clients... needs cleanup 👍

Reply:
Restructured - surprised this wasn't a lint error in pedantic mode here!